iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp

```diff
@@ -0,0 +1,139 @@
+#include "iceoryx_utils/cxx/optional.hpp"
+#include "test.hpp"
+
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+class GenericRAII_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(GenericRAII_test, InitFunctionIsCalledInCtorWhenSet)
+{
+    int hasCalledInit = 0;
+    GenericRAII sut([&] { ++hasCalledInit; }, std::function<void()>());
+    EXPECT_THAT(hasCalledInit, Eq(1));
+}
+
+TEST_F(GenericRAII_test, InitFunctionIsCalledInCtorWhenSetWithCleanupFunction)
+{
+    int hasCalledInit = 0;
+    int hasCalledCleanup = 0;
+    GenericRAII sut([&] { ++hasCalledInit; }, [&] { ++hasCalledCleanup; });
+    EXPECT_THAT(hasCalledInit, Eq(1));
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtor)
+{
+    int hasCalledInit = 0;
+    int hasCalledCleanup = 0;
+
+    {
+        GenericRAII sut([&] { ++hasCalledInit; }, [&] { ++hasCalledCleanup; });
+    }
+
+    EXPECT_THAT(hasCalledInit, Eq(1));
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtorWhenUsingCleanupOnlyCTor)
+{
+    int hasCalledCleanup = 0;
+
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+    }
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
+
+TEST_F(GenericRAII_test, CleanupFunctionIsCalledInDtorWithEmptyInitFunction)
+{
+    int hasCalledCleanup = 0;
+
+    {
+        GenericRAII sut(function_ref<void()>(), [&] { ++hasCalledCleanup; });
+    }
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
+
+TEST_F(GenericRAII_test, MoveCTorDoesNotCallCleanupFunctionOfOrigin)
+{
+    int hasCalledCleanup = 0;
+
+    GenericRAII sut([&] { ++hasCalledCleanup; });
+    GenericRAII sut2(std::move(sut));
+
+    EXPECT_THAT(hasCalledCleanup, Eq(0));
+}
+
+TEST_F(GenericRAII_test, MoveConstructedDoesCallCleanupFunctionWhenDestroyed)
+{
+    int hasCalledCleanup = 0;
+
+    {
+        iox::cxx::optional<GenericRAII> sut(GenericRAII([&] { ++hasCalledCleanup; }));
+        GenericRAII sut2(std::move(*sut));
+        sut.reset();
+        EXPECT_THAT(hasCalledCleanup, Eq(0));
+    }
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
+
+TEST_F(GenericRAII_test, MoveAssignmentCallsCleanup)
+{
+    int hasCalledCleanup = 0;
+    int hasCalledCleanup2 = 0;
+
+    GenericRAII sut([&] { ++hasCalledCleanup; });
+    GenericRAII sut2([&] { ++hasCalledCleanup2; });
+
+    sut = std::move(sut2);
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+    EXPECT_THAT(hasCalledCleanup2, Eq(0));
+}
+
+TEST_F(GenericRAII_test, MoveAssignedCallsCleanupWhenOutOfScope)
+{
+    int hasCalledCleanup = 0;
+    int hasCalledCleanup2 = 0;
+
+    {
+        GenericRAII sut([&] { ++hasCalledCleanup; });
+        GenericRAII sut2([&] { ++hasCalledCleanup2; });
+
+        sut = std::move(sut2);
+        EXPECT_THAT(hasCalledCleanup, Eq(1));
+        EXPECT_THAT(hasCalledCleanup2, Eq(0));
+    }
+
+    EXPECT_THAT(hasCalledCleanup2, Eq(1));
```

Comment:
I'm wondering if `hasCalledCleanup` should also be checked here. Technically it's done with the move ctor due to code reuse, but that's just an implementation detail.